### PR TITLE
[v12] Require SSH prefix when proxying connection to a host

### PR DESCRIPTION
**Pull request by AntonAM**
Logical backport https://github.com/gravitational/teleport/pull/33000 to branch/v12

It's not identical backport - I returned to first version of the original PR with checking read side - we can't put restriction on the write side inside router.DialHost() because there's a chance we'll do handshake with ProxyHelloSignature for old nodes.

### CODE REVIEW

**Review by jentfoo — APPROVED** [review]

**Review by espadolini — DISMISSED** [review]
This will make a new bufio reader on top of the inner RWC on every read, reading (in general) more than what will be returned to the user, then discarding the additional data.

**Review by espadolini — COMMENTED** [review]
I'd rather this use the same general approach as the main PR, without adding an additional buffer in the middle; we can just passively check the bytes read from the upstream reader, right?

**Comment by AntonAM** [comment]
Changed implementation to similar approach to the one in the main PR.

**Review by jentfoo — APPROVED** [review]

**Review by espadolini — APPROVED** [review]

File: lib/srv/regular/proxy.go [hunk 1]
```diff
@@ -270,12 +270,55 @@ func (t *proxySubsys) proxyToHost(ctx context.Context, ch ssh.Channel, clientSrc
...
+
+	// Decide which is smaller, read data or remaining portion of the required prefix
+	small, big := c.requiredPrefix[c.requiredPointer:], b[:n]
+	if len(small) > len(big) {
+		big, small = small, big
+	}
+
+	if !bytes.HasPrefix(big, small) {
+		c.requiredPointer = -1
+		return n, trace.AccessDenied("required prefix %q was not found", c.requiredPrefix)
```
Comment (espadolini) [inline lib/srv/regular/proxy.go:313]: We should not allow the invalid data "out" of the function (idk if we need to clear the data from the buffer, I reckon it's fine to just say it's unspecified).

```suggestion
		return 0, trace.AccessDenied("required prefix %q was not found", c.requiredPrefix)
```

**Review by rosstimothy — APPROVED** [review]

**Pull request merged**